File: src/main/java/com/diluv/api/utils/ResponseUtilities.java
package com.diluv.api.utils;

import com.diluv.api.error.ErrorMessages;
import com.diluv.api.error.Errors;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.core.json.Json;
import io.vertx.ext.web.RoutingContext;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ResponseUtilities {

    public static HttpServerResponse asJSONResponse(RoutingContext route, int statusCode, Map<String, Object> additionalData, Object data) {
        HttpServerResponse response = route.response();
        response.setStatusCode(statusCode);
        response.putHeader("Content-Type", "application/json; charset=utf-8");
        response.putHeader("Access-Control-Allow-Origin", "*");
        response.putHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        response.putHeader("Access-Control-Allow-Headers", "Content-Type, Authorization");

        Map<String, Object> out = new HashMap<>();
        out.put("statusCode", statusCode);
        if (data != null)
            out.put("data", data);
        if (additionalData != null)
            out.putAll(additionalData);
        response.end(Json.encodePrettily(out));
        return response;
    }

    public static HttpServerResponse asErrorResponse(RoutingContext route, Errors error, Object additionalData) {

        Map<String, Object> map = new HashMap<>();
        map.put("error", error.errorMessage);

        if (additionalData != null)
            map.put("message", additionalData);
        return ResponseUtilities.asJSONResponse(route, error.statusCode, map, null);
    }

    public static HttpServerResponse asErrorResponse(RoutingContext route, ErrorMessages errorMessages) {
        return ResponseUtilities.asErrorResponse(route, errorMessages.errors, errorMessages.errorMessage);
    }

    public static HttpServerResponse asErrorResponse(RoutingContext route, Errors error, List<ErrorMessages> errorMessagesList) {
        List<String> errorMessages = new ArrayList<>();
        for(ErrorMessages errorMessage:errorMessagesList){
            errorMessages.add(errorMessage.errorMessage);
        }
        return ResponseUtilities.asErrorResponse(route, error, errorMessages);
    }

    public static HttpServerResponse asSuccessResponse(RoutingContext route, Map<String, Object> data) {
        return ResponseUtilities.asJSONResponse(route, 200, null, data);
    }

    public static HttpServerResponse asSuccessResponse(RoutingContext route, List data) {
        return ResponseUtilities.asJSONResponse(route, 200, null, data);
    }
}

File: src/main/kotlin/com/diluv/api/utils/ResponseUtilities.kt
package com.diluv.api.utils;

import com.diluv.api.error.Errors
import io.vertx.core.http.HttpServerResponse
import io.vertx.core.json.Json
import io.vertx.ext.web.RoutingContext

fun RoutingContext.asJSONResponse(statusCode: Int, additionData: Map<String, Any>?, data: Any?): HttpServerResponse {
    val response = this.response()
    response.statusCode = statusCode
    response.putHeader("Content-Type", "application/json; charset=utf-8")
    response.putHeader("Access-Control-Allow-Origin", "*")
    response.putHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
    response.putHeader("Access-Control-Allow-Headers", "Content-Type, Authorization")
    var out = mapOf<String, Any>("statusCode" to statusCode)
    if (data != null)
        out += mapOf("data" to data)
    if (additionData != null)
        out += additionData
    response.end(Json.encodePrettily(out))
    return response
}

fun RoutingContext.asErrorResponse(error: Errors, additionData: Any?): HttpServerResponse {
    var map = mapOf<String, Any>("error" to error.errorMessage)
    if (additionData != null)
        map += mapOf("message" to additionData)
    return this.asJSONResponse(error.statusCode, map, null)
}

fun RoutingContext.asSuccessResponse(data: Map<String, Any>): HttpServerResponse {
    return this.asJSONResponse(200, null, data)
}

fun RoutingContext.asSuccessResponse(data: List<Any>): HttpServerResponse {
    return this.asJSONResponse(200, null, data)
}